File: 2024/day10/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = false;
const EXAMPLE = false;

class Coord {
  constructor(public row: number, public col: number) {}
  hash(): number {
    return this.row * 10000000 + this.col;
  }
  validate(maxRow: number, maxCol: number): boolean {
    return this.row >= 0 && this.row < maxRow && this.col >= 0 && this.col < maxCol;
  }
}

const text = await Deno.readTextFile(EXAMPLE ? 'example.txt' : 'input.txt');
if (DEBUG) console.debug('text', text);

const mapData = text
  .split('\n')
  .filter((line) => line.trim() != '')
  .map((line) => line.split('').map((char) => parseInt(char)));
const countRows = mapData.length;
const countCols = mapData[0].length;
if (DEBUG) console.debug({ mapData, countRows, countCols });

const trailheads = mapData.reduce<Coord[]>((rowAcc, rowData, rowIx) => {
  const rowMatches = rowData.reduce<Coord[]>((colAcc, value, colIx) => {
    if (value == 0) {
      colAcc.push(new Coord(rowIx, colIx));
    }
    return colAcc;
  }, []);
  return rowAcc.concat(rowMatches);
}, []);
if (DEBUG) console.debug({ trailheads, len: trailheads.length });

let trails: Map<number, Coord>[] = trailheads.map((trailhead) => new Map([[trailhead.hash(), trailhead]]));
if (DEBUG) console.debug({ trails });

const trailsCompleted: Map<number, Coord>[] = [];
for (let nextValue = 1; nextValue < 10; nextValue++) {
  if (DEBUG) console.debug({ nextValue, trails: trails.length, completed: trailsCompleted.length });
  const trailsNext: Map<number, Coord>[] = [];
  for (const trail of trails) {
    // const coord = trail[nextValue - 1];
    const trailValues = [...trail.values()];
    const coord = trailValues.pop()!;
    for (let dirIx = 0; dirIx < 4; dirIx++) {
      const coordNext = new Coord(
        coord.row + (dirIx == 0 ? -1 : dirIx == 2 ? 1 : 0),
        coord.col + (dirIx == 3 ? -1 : dirIx == 1 ? 1 : 0)
      );
      // if (DEBUG) console.debug('test', { coord, dirIx, coordNext });
      if (
        coordNext.validate(countRows, countCols) &&
        !trail.has(coordNext.hash()) &&
        mapData[coordNext.row][coordNext.col] == nextValue
      ) {
        if (DEBUG) console.debug('move', { coord, coordNext, dirIx, nextValue });
        const trailNext = new Map([...trailValues, coord, coordNext].map((i) => [i.hash(), i]));
        if (nextValue == 9) {
          trailsCompleted.push(trailNext);
        } else {
          trailsNext.push(trailNext);
        }
      }
    }
  }
  trails = trailsNext;
}
const trailStartEnd = new Set(
  trailsCompleted.map((trail) => {
    const values = [...trail.values()];
    return JSON.stringify([values[0], values[9]]);
  })
);
if (DEBUG) console.debug({ trailStartEnd });
const totalScores = trailStartEnd.size;
console.log('part 1', totalScores);

const totalRating = trailsCompleted.length;
console.log('part 2', totalRating);

File: 2024/day10/solution.py
#!/usr/bin/env python3

DEBUG = True
EXAMPLE = False

with open('example.txt' if EXAMPLE else 'input.txt', 'r') as f:
  heights = [list(map(int, x)) for x in f.read().splitlines()]
if DEBUG: print(f'{heights=}')

count_rows = len(heights)
count_cols = len(heights[0])
trailheads: list[tuple[int, int]] = []
for row_ix in range(count_rows):
  for col_ix in range(count_cols):
    if heights[row_ix][col_ix] == 0:
      trailheads.append((row_ix, col_ix))
if DEBUG: print(f'{trailheads=}')
trails = [[x] for x in trailheads]
if DEBUG: print(f'{trails=}')

trails_completed: list[list[tuple[int, int]]] = []
trails_next: list[list[tuple[int, int]]] = []
found = False
while not (found):
  for trail in trails:
    coord = trail[-1]
    value = heights[coord[0]][coord[1]]
    # if DEBUG: print(f'{trail=} {coord=} {value=}')
    value_next = value + 1
    for direction_ix in range(4):
      coord_next = (
        coord[0] + (-1 if direction_ix == 0 else 1 if direction_ix == 2 else 0),
        coord[1] + (-1 if direction_ix == 3 else 1 if direction_ix == 1 else 0),
      )
      if 0 <= coord_next[0] < count_rows and 0 <= coord_next[1] < count_cols and \
        coord_next not in trail and \
        heights[coord_next[0]][coord_next[1]] == value_next:
        trail_next = [*trail, coord_next]
        assert 0 < value_next < 10
        if value_next == 9:
          if DEBUG: print(f'completed {trail_next}')
          trails_completed.append(trail_next)
          found = True
        else:
          assert len(trail_next) < 10
          trails_next.append(trail_next)

  print(f'{len(trails)=} {len(trails_next)=} {len(trails_completed)=}')
  trails = trails_next

# completed_trails could have multiple routes with the same dest and origin
trails_completed_set = {(x[0], x[-1]) for x in trails_completed}
total_scores = len(trails_completed_set)
print(f'part 1 {total_scores}')

total_ratings = len(trails_completed)
print(f'part 2 {total_ratings}')
